src/components/structures/RoomDirectory.tsx

```diff
@@ -116,7 +119,30 @@ export default class RoomDirectory extends React.Component<IProps, IState> {
         } else if (!selectedCommunityId) {
             MatrixClientPeg.get().getThirdpartyProtocols().then((response) => {
                 this.protocols = response;
-                this.setState({ protocolsLoading: false });
+                const myHomeserver = MatrixClientPeg.getHomeserverName();
+                const lsRoomServer = localStorage.getItem(LAST_SERVER_KEY);
+                const lsInstanceId = localStorage.getItem(LAST_INSTANCE_KEY);
+                const configSevers = SdkConfig.get().roomDirectory?.servers || [];
+                const settingsServers = SettingsStore.getValue("room_directory_servers") || [];
+                const roomServer = [...configSevers, ...settingsServers].includes(lsRoomServer)
+                    ? lsRoomServer
+                    : myHomeserver;
+                const instanceIds = [];
+                if (roomServer === myHomeserver) {
+                    Object.values(this.protocols).forEach((protocol) => {
+                        protocol.instances.forEach((instance) => instanceIds.push(instance.instance_id));
+                    });
+                }
+                const instanceId = (instanceIds.includes(lsInstanceId) || lsInstanceId === ALL_ROOMS)
+                    ? lsInstanceId
+                    : null;
+
+                this.setState({
+                    protocolsLoading: false,
+                    instanceId: instanceId,
+                    roomServer: roomServer,
+                });
+                this.refreshRoomList();
```

Comment:
this is a bit annoying, could we not assume that the LS values are correct, load them in c'tor and then just validate them here, triggering a refresh only if the validation failed and they had to be reset